salt/modules/system.py

```diff
@@ -608,3 +608,54 @@ def get_computer_name():
         salt '*' network.get_hostname
     '''
     return __salt__['network.get_hostname']()
+
+
+NILRT_REBOOT_WITNESS_PATH = '/var/volatile/tmp/salt/reboot_witnessed'
+
+
+def set_reboot_required_witnessed():
+    '''
+    This function is used to remember that an event indicating that a reboot is
+    required was witnessed. This function writes to a temporary filesystem so
+    the event gets cleared upon reboot.
+
+    Returns:
+        bool: ``True`` if successful, otherwise ``False``
+
+    .. code-block:: bash
+
+        salt '*' system.set_reboot_required_witnessed
+    '''
+    if __grains__.get('os_family') == 'NILinuxRT':
```

Comment:
Hrm. I don't like that this function is essentially a no-op for everything _but_ the `NILinuxRT` OS family but exists in the `system` module which is available and consumed by all types of systems. Is this the best place for this?

Reply:
@cachedout : I think the idea behind this was to copy from the architecture already put in place in `salt/modules/win_system.py`, which is also a "system" module (by virtual name). So the function names and location are not new. I believe this exact same feature could apply to other Linux distributions, I would assume that the author was not motivated to implement and test it on other distributions. Although, by looking at things, it seems that it wouldn't be difficult to do so.

Perhaps the author should add at least one more distribution (eg CentOS) so that your automated tests would cover this new feature?

Reply:
I can move it somewhere else, no problem, we just need to figure out where is the best place to put them. How about the opkg module? It was written for the NILinuxRT family and it's the only place which calls these witness functions.

Reply:
I think that opkg.py is probably the correct place to put it. I don't see any code here that actually calls these new functions, and since they only apply to NILinuxRT they shouldn't be in a more general module.

Reply:
@terminalmage : I disagree, because I think these new functions are not private and really meant to be called at the job level. Eg `system.get_reboot_required_witnessed`. How else would the master find out if a reboot is required on a given minion? Because they are required to be called at the job level, putting them into a module with virtual name `pkg` doesn't make that much sense, because all modules with virtual name `pkg` should adhere to a predefined set of APIs, which this falls outside of.

Since there already is precedent for `system.get_reboot_required_witnessed` from the Windows side, it seems like the logical place to put these functions where they are right now. This feature can easily be extended to other Linux distributions, because the main logic that checks if a reboot is required is in `salt/modules/restartcheck.py`, which already supports Debian and Red Hat based systems.

The `pkg` modules for such systems could be modified in a similar way to check if a restart is required after a package operation, and invoke `system.set_reboot_required_witnessed` in such a case. Then, for such systems, a user has the option of checking if a reboot is required after a package operation (eg if a new kernel is installed). It would actually be a pretty useful feature for general use.

Reply:
I'm ok either way, I think we just need a consensus so I know what to do further.

Reply:
@skizunov Thanks, I overlooked your earlier comment regarding the `win_system` module. Sorry about that! I agree, system.py is a good place for it.